opal/accelerator: Initial accelerator framework implementation

wckzhang opened the pull request:
Add new accelerator framework. Initial assumption is that
only one accelerator component will be selected per node.

Added a cuda component
Added a null component (no devices, performs no-ops/errors)
Replaced a large chunk of common_cuda usage (it's not all unused, but it's mostly unused).
Most likely broke ROCM usage until the ROCM component is written

Signed-off-by: William Zhang <wilzhang@amazon.com>

ibm-ompi commented:
The IBM CI (GNU/Scale) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/45dd1add3d5625710fadc0c4315ac2fe

ibm-ompi commented:
The IBM CI (XL) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/6e2893f0ac301fcd63b998d1e888efca

ibm-ompi commented:
The IBM CI (PGI) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/96d9cf89dde7cf1eba0c47721777ae58

edgargabriel commented on opal/mca/accelerator/accelerator.h:
```diff
@@ -0,0 +1,314 @@
+    opal_accelerator_base_module_open_handle_fn_t open_handle;
+    opal_accelerator_base_module_close_handle_fn_t close_handle;
+
+    opal_accelerator_base_module_host_unregister_fn_t host_register;
```
probably just a type, should this pointer be of type ``` opal_accelerator_base_module_host_register_fn_t```

wckzhang replied:
Thanks

bosilca commented:
:+1: I have a proposal, let's drop the memcpy functions and use a vector_copy instead, defined as (count, blen, stride). This is the most basic type in the datatype, and any contiguous memory can easily be translated into this (stride = blen or something). For all the accelerators that support vector copy (e.g. cudaMemcpy2D) there will be a significant gain in performance. For everybody else it will not really matter, the vector_copy can simply do a loop around a memcpy.

edgargabriel commented:
@bosilca, I like the idea, but I am wandering whether we will need maybe both, a regular memcpy and a vector_copy operation. I am thinking about integration with e.g. UCX, not sure how easy a change to vector_copy on that level would be.

bwbarrett commented:
Edgar, isn't it just manually unroll the vector copy into single memcopys?

edgargabriel commented:
@bwbarrett I am more worried about the fact that the vector_copy will have different signature than the memcpy function, and if e.g. UCX has to be adjusted to use this interface, it won't be able to be compiled with a previous version of ompi.

edgargabriel commented:
actually, I take it back, maybe its not an issue. They are invoking the opal_convertor_pack/unpack, and that interface would probably not change.

bwbarrett commented on opal/mca/accelerator/Makefile.am:
```diff
@@ -0,0 +1,34 @@
+# $HEADER$
+#
+
+AM_CPPFLAGS = $(LTDLINCL) $(accelerator_CPPFLAGS)
```
You don't need this line at all.  You don't need the `$(LTDLINCL)` because we don't use libltdl for its faked dlopen support.  you don't need `$(accelerator_CPPFLAGS)` because you don't have a framework-level configure.m4 snippet to ever set `$(accelerator_CPPFLAGS)`.

a-szegel replied:
Line Removed.

bwbarrett commented on opal/mca/accelerator/accelerator.h:
```diff
@@ -0,0 +1,295 @@
+/**
+ * Accelerator flags
+ */
+#define MCA_ACCELERATOR_UNIFIED_MEMORY 0x00000001
```
please add documentation of what this flag means.  If these are flags to be used in the accelerator `flags` fields throughout this header, please add `FLAGS` somewhere in this name.

a-szegel replied:
Updated with:
```
/**
 * Accelerator flags
 */
/* Unified memory buffers */
#define MCA_ACCELERATOR_FLAGS_UNIFIED_MEMORY 0x00000001
```

bwbarrett commented on opal/mca/accelerator/accelerator.h:
```diff
@@ -0,0 +1,295 @@
+ *                           about the corresponding pointer such
+ *                           as whether it belongs to unified memory.
+ *
+ * @return <0                An error has occurred.
```
This isn't proper doxygen format.  `@retval` specifies a return value and description.  `@return` is a paragraph form describing the return value.

Does this return OPAL error codes or some other error scheme?

a-szegel replied:
verified Doxygen comments have been updated.

bwbarrett commented on opal/mca/accelerator/accelerator.h:
```diff
@@ -0,0 +1,295 @@
+ *
+ * @return                   OPAL_SUCCESS or error status on failure
+ */
+typedef int (*opal_accelerator_base_module_get_device_fn_t)(
```
I thought you said that Ze didn't have this assumption and could use multiple devices in a single process?

bwbarrett replied:
So you're maintaining the assumption that a single process "controls" at most one accelerator.  Does that match with what Intel and/or AMD want?

edgargabriel replied:
Assuming this would be implemented using cudaGetDevice/hipGetDevice, in all reality it would return which ever device has been set by the user in the code, unless you want to cache internally the value. The AMD ROCr runtime library does not have the notion of a 'current' device, it is purely maintained on the hip layer.

bwbarrett replied:
Did we ever sort this out?

wckzhang replied:
Added device ids back in after a conversation with Sean Hefty from Intel

bwbarrett commented on opal/mca/accelerator/base/accelerator_base_select.c:
```diff
@@ -0,0 +1,148 @@
+
+    /* Finished querying all components. Check for no components found */
+
+    if( NULL == best_component ) {
```
why is this an error?  It's also pretty weird that you call show_help, but then continue on through the function.

wckzhang replied:
Returned an error code

wckzhang replied:
I imagined any component trying to use the accelerator framework will read that it isn't OPAL_SUCCESS (or OPAL_ERR_NOT_FOUND) and will proceed without using the accelerator framework.  I suppose if we move toward having a list of accelerators, it would just be an empty list and any logic should just work with an empty list.

bwbarrett replied:
I can't think of an example of a non-required framework's select returning an error code if no components are found, but I could be wrong.  Obviously, something like the PML would puke in that situation.  But the onesided and btl componetns would continue with empty lists (until something required a component in that framework, then pukage).  I think the empty list is the best approach; works for both today and any potential multi-component future.

bwbarrett commented on opal/mca/accelerator/base/accelerator_base_select.c:
```diff
@@ -0,0 +1,148 @@
+                       "accelerator");
+    }
+
+    opal_output_verbose( 10, opal_accelerator_base_framework.framework_output,
```
be consistent in your files with calling functions as `func()` vs. `func ()`.  Everyone but Nathan appears to prefer `func()`, which is what I would like to see in this PR.

wckzhang replied:
Fixed

bwbarrett commented on opal/mca/accelerator/base/accelerator_base_select.c:
```diff
@@ -0,0 +1,148 @@
+} accelerator_opened_component_t;
+
+
+int opal_accelerator_base_select(void)
```
I question whether this is the right selection logic.  We're saying that there's a "best" accelerator, which feels like a path to madness today.  I think there are two valid options for the selection logic:

1. We can have at most 1 active accelerator framework.  If we find multiple that say they work, we give up and make the user pick.  The at most 1 is important; 0 feels like a fairly normal case.
2. We can support multiple simultaneous accelerator components.

bwbarrett commented:
While I understand the desire for comments, we shouldn't merge this PR until 1) the code changes to actually use it are in place and 2) there's at least one component.  I'm not sure we can really be confident in the interface until we're at that point.

wckzhang commented:
Need to:
Rework component selection logic
Replace code that calls common_cuda that can call the framework device agnostically instead

wckzhang commented:
Looks like the PR build checker failed due to jenkins losing connection, will fix itself on a re-run

wckzhang commented:
I changed the datatype engine, mtl's, common/ompio, osc/rdma, and mtls to use the new framework. The remaining dependencies on common_cuda.h are in:

- coll_cuda component
- nbc_internal
- pml ob1
- pml ucx
- btl smcuda
- btl tcp
- rcache gpusm module

I don't plan to change (for now)
- coll_cuda component
- rcache gpusm module
- pml ucx (Without feedback from someone with stake in ucx)

I think I'll remove the dependency in nbc_internal before I merge the patch. The remaining changes are more in depth changes in the ob1/btl layer.

bwbarrett commented:
I think you need to fix nbc, pb1, and tcp before merging (ie, all the non-cuda-specific components).  It's hard is precisely the reason to do the changes.  We need to be 100% sure that we have the right interface.

bwbarrett commented on ompi/mca/common/ompio/common_ompio_buffer.c:
```diff
@@ -39,26 +39,29 @@ static int32_t  mca_common_ompio_pagesize=4096;
     
-    convertor.flags=0;
-    if ( opal_cuda_check_one_buf ( (char *)buf, &convertor ) ) {
+    if (NULL == opal_accelerator.check_addr) {
```
Why not have base functions that are essentially no-ops, so you don't need this check?  This is a performance part of the code, and we're adding unnecessary conditionals.  I'm only going to make this comment here, but it applies throughout the PR.

wckzhang replied:
Noted, will change

bwbarrett commented on ompi/mca/mtl/base/mtl_base_datatype.h:
```diff
@@ -26,92 +26,27 @@
 						  convertor->count)) {
-	    *freeAfter = false;
-	    *buffer = convertor->pBaseBuf;
+	    *freeAfter = false; *buffer = convertor->pBaseBuf;
```
this shouldn't be on one line.

a-szegel replied:
fixed.

bwbarrett commented on ompi/mca/mtl/base/mtl_base_datatype.h:
```diff
@@ -125,138 +60,96 @@ ompi_mtl_datatype_pack(struct opal_convertor_t *convertor,
+    /* opal_convertor_need_buffers always returns true
+     * if CONVERTOR_ACCELERATOR is set, so unset temporarily
+     */
+    convertor->flags &= ~CONVERTOR_ACCELERATOR;
```
This feels like the wrong approach to me.  We shouldn't go mucking with convertor flags because of an internal behavior of the converter interface; we should fix the convertor interface.

wckzhang replied:
I'll rethink this

wckzhang replied:
Fixed this

bwbarrett commented on ompi/mca/mtl/ofi/mtl_ofi_component.c:
```diff
@@ -543,7 +537,7 @@ static int ompi_mtl_ofi_init_regular_ep(struct fi_info * prov, int universe_size
             }
     }
-#endif /* OPAL_CUDA_SUPPORT && FI_OPT_FI_HMEM_P2P */
+#endif /* HAVE_DECL_FI_OPT_FI_HMEM_P2P */
```
For lack of a better place to leave this note, you need to do something with the OFI MTL's configure.m4 as well.

bwbarrett replied:
This note appears to still apply.

wckzhang replied:
Done

bwbarrett commented on ompi/mca/mtl/ofi/mtl_ofi_component.c:
```diff
@@ -802,14 +791,12 @@ ompi_mtl_ofi_component_init(bool enable_progress_threads,
 
-#if OPAL_CUDA_SUPPORT
     if (!(prov->caps & FI_HMEM)) {
         opal_output_verbose(1, opal_common_ofi.output,
```
This printf is now going to always fire for providers that don't support FI_HMEM.  That's a fairly big change in behavior.  More thought is needed here.

bwbarrett replied:
This note still applies.

wckzhang replied:
Changing behavior to:

Try to select a provider with FI_HMEM support. If this fails, proceed selecting provider w/o FI_HMEM support anyway but set a bool that declares accelerator transfers aren't supported. Change MTL components to report back if they support directly sending using accelerator buffers. If not, copy from accelerator to host buffers.

bwbarrett commented on ompi/mca/pml/ob1/pml_ob1_recvfrag.c:
```diff
@@ -787,22 +776,6 @@ void mca_pml_ob1_recv_frag_callback_frag (mca_btl_base_module_t *btl,
     recvreq = (mca_pml_ob1_recv_request_t*)hdr->hdr_frag.hdr_dst_req.pval;
-#if OPAL_CUDA_SUPPORT /* CUDA_ASYNC_RECV */
-    /* If data is destined for GPU buffer and convertor was set up for asynchronous
-     * copies, then start the copy and return.  The copy completion will trigger
```
where did this code go?

bwbarrett reviewed (commented):
It looks like you removed all the CUDA stream code.  Are we sure that's ok?

a-szegel commented on opal/mca/accelerator/cuda/accelerator_cuda.c:
```diff
@@ -0,0 +1,775 @@
+    return OPAL_SUCCESS;
+}
+
+opal_accelerator_base_module_t* accelerator_cuda_internal_init(void)
```
This init function should fail if:

1. The user did not initialize a CUDA context before this function got called
2. There is not a CUDA device present on the instance

I think we need to make a call into the CUDA API in this method to confirm that the CUDA component we are returning is functional.  If it is not functional, we should return the Null Accelerator to the user.

a-szegel replied:
Fixed!

a-szegel commented on opal/mca/accelerator/cuda/accelerator_cuda.c:
```diff
@@ -0,0 +1,775 @@
+    return OPAL_SUCCESS;
+}
+
+opal_accelerator_base_module_t* accelerator_cuda_internal_init(void)
```
This init function should fail if:

1. The user did not initialize a CUDA context before this function got called
2. There is not a CUDA device present on the instance

I think we need to make a call into the CUDA API in this method to confirm that the CUDA component we are returning is functional.  If it is not functional, we should return the Null Accelerator to the user.

wckzhang commented:
Added new asynchronous API's, Seth added a null component, rewrote the asynchronous progress engine in the ob1 pml, replaced all the OPAL_CUDA_SUPPORT ifdefs in the pml and btls (non cuda).

I'm pretty sure that things have broken for ROCM since a lot of conflicts appeared during the rebase. I didn't try fixing them as they will get addressed when Edgar writes the ROCM component.

bwbarrett commented on config/opal_check_cuda.m4:
```diff
@@ -90,7 +90,7 @@ AS_IF([test "$opal_check_cuda_happy" = "yes"],
 # If we have CUDA support, check to see if we have support for SYNC_MEMOPS
 # which was first introduced in CUDA 6.0.
-AS_IF([test "$opal_check_cuda_happy"="yes"],
+AS_IF([test "$opal_check_cuda_happy" = "yes"],
```
The syntax error check PR appears to already be in main.  Did you mis-rebase somewhere?

wckzhang replied:
Let me check

edgargabriel commented on ompi/mca/osc/rdma/osc_rdma_component.c:
```diff
@@ -378,14 +376,14 @@ static int ompi_osc_rdma_component_query (struct ompi_win_t *win, void **base, s
-        if (opal_cuda_check_bufs(*base, NULL)) {
+        int dev_id;
+        uint64_t flags;
+        if (opal_accelerator.check_addr(*base, &dev_id, &flags)) {
```
I think the real question for this component (and some other components in ompi) is not necessarily 'is it a device buffer', but 'can we execute operations x/y/z on this address?' This could be achieved using the flags for the check_addr() function by defining the corresponding flags in accelerator.h, or we could introduce a different interface (aka check_capabilities() or similar.)

wckzhang replied:
This was sorted out offline through slack

edgargabriel commented on opal/datatype/opal_datatype_pack.h:
```diff
@@ -106,8 +106,8 @@ static inline void pack_predefined_data(opal_convertor_t *CONVERTOR, const dt_el
-             !(CONVERTOR->flags & CONVERTOR_ROCM)
+        if (!(CONVERTOR->flags & CONVERTOR_ACCELERATOR) &&
+            !(CONVERTOR->flags & CONVERTOR_ROCM)
             && OPAL_LIKELY(
```
Do we really need the CONVERTOR->flags & CONVERTOR_ROCM or is it just a stop-gap measure for now? (same question applies to a couple of other places in opal/datatype).

wckzhang replied:
Stop gap

bwbarrett replied:
ew; this needs to go.  We shouldn't have a stop-gap like this.

wckzhang commented on opal/datatype/opal_datatype_copy.c:
```diff
@@ -67,78 +55,104 @@ static size_t opal_datatype_memop_block_size = 128 * 1024;
+        dev_id = 0;
+    }
+    res = opal_accelerator.memcpy(dev_id, dest, src, size, type);
+    if (OPAL_SUCCESS != res) {
```
I'm thinking it's not really that useful. I added that a while ago because of the way ZE memcpy works...but there's probably a way for the implementation to handle it. Since the libfabric API had a dev_id, I figured that it was there mainly for ZE

wckzhang replied:
Added it to give more flexibility in the future, but also allowed -1 (no device id) as a valid input

wckzhang commented on opal/datatype/opal_convertor.h:
```diff
@@ -35,31 +35,31 @@
+#define CONVERTOR_NO_OP                  0x00100000
+#define CONVERTOR_WITH_CHECKSUM          0x00200000
+#define CONVERTOR_ACCELERATOR            0x00400000
+#define CONVERTOR_ROCM                   0x00400000 //same as CUDA on purpose
```
Yeah, we're going with the assumption that there will only be one type of accelerator, and this should be decided from the accelerator selection logic. Thus we only need to flag it as a GPU buffer of some type.

bwbarrett replied:
This is really weird; why are we not adding ROCM components and kiling off this constant?

wckzhang replied:
We are going to add a ROCM component, Edgar was going to work on it against this PR.

wckzhang commented on opal/datatype/opal_datatype_copy.c:
```diff
@@ -67,78 +55,104 @@ static size_t opal_datatype_memop_block_size = 128 * 1024;
+        dev_id = 0;
+    }
+    res = opal_accelerator.memmove(dev_id, dest, src, size, type);
+    if (OPAL_SUCCESS != res) {
```
Answered above

a-szegel commented on opal/mca/accelerator/accelerator.h:
```diff
@@ -0,0 +1,480 @@
+    MCA_ACCELERATOR_TRANSFER_DTOD,
+} opal_accelerator_transfer_type_t;
+
+struct opal_accelerator_stream_t {
```
I don't understand why you created `opal_accelerator_stream_t` and `opal_accelerator_event_t` structures?  Why don't you just pass these as void pointers in the interface?  Why create a structure around a void pointer, what does that get you?

wckzhang replied:
This was actually a request from Brian. What I assume this allows is for more flexibility for the implementer (ie, they can add fields to the object as needed)

bwbarrett replied:
I think I missed this because it was resolved by someone, but I wasn't asking for a named structure with a single void* inside of it.  I was asking that we avoid void* in our APIs, because it's hard to figure out object types in either side of the code, and very easy to feed the wrong type of object.

But I was also asking that streams and events get made into OMPI objects, which means that the user of the API has a nice, opaque object it can use, and each component can have a customized storage for streams and events (that match their implementations) by inheriting from the base stream/event structures.  So you have an abstract object for callers that can be used for data storage by the implementations.  But this requires making these proper OMPI objects, which also changes your API (as mentioned in comments later in this file).

But, to be clear, I see no value in a structure with a single member that is a void*.  If you're going to do that, just typedef opal_accelerator_stream_t to be a void*.

a-szegel commented on opal/mca/accelerator/accelerator.h:
```diff
@@ -0,0 +1,480 @@
+ * @return                   OPAL_SUCCESS or error status on failure.
+ */
+typedef int (*opal_accelerator_base_module_create_event_fn_t)(
+    opal_accelerator_event_t *event);
```
I don't understand why `create_stream` takes a `opal_accelerator_stream_t **stream` and `create_event` takes an `opal_accelerator_event_t *event`.  This implies that the Accelerator API will fully allocate a stream, but require an event to be partially pre-allocated (the create_stream API still allocates).  This causes the user to treat them differently, when they are fundamentally the same thing (a structure with 1 void pointer inside):

```
    opal_accelerator_stream_t *stream = NULL;
    opal_accelerator.create_stream(dev_id, &stream)
    
    opal_accelerator_event_t event = {};
    opal_accelerator.create_event(&event))
```

wckzhang replied:
Yeah I might change create_stream to not allocate. The reason I changed create_event to not allocate is because of the asynchronous progress engine. The design wanted to allocate a large block for a circular event array which wouldn't have been possible with create_event doing the allocation.

bwbarrett commented on opal/datatype/opal_convertor.c:
```diff
@@ -40,18 +40,33 @@
-#    include "opal/cuda/common_cuda.h"
-#    define MEMCPY_GPU(DST, SRC, BLENGTH, CONVERTOR) \
+#include "opal/mca/accelerator/accelerator.h"
+#    define MEMCPY_ACCELERATOR(DST, SRC, BLENGTH, CONVERTOR) \
```
blank line between the includes and the define.  Also, with the #ifdef removed, there shouldn't be spacing between the `#` and `define`.

a-szegel replied:
Cleaned up.

bwbarrett commented on opal/datatype/opal_convertor.c:
```diff
@@ -40,18 +40,33 @@
+        /**
+         * Attempt the copy with 0 as dev_id. If we fail the copy then we'll abort anyway.
+         */
+        dev_id = 0;
```
This makes no sense.  Why is this the right semantic?

bwbarrett commented on opal/datatype/opal_convertor.c:
```diff
@@ -60,9 +75,7 @@ static void opal_convertor_construct(opal_convertor_t *convertor)
-    convertor->cbmemcpy = &opal_cuda_memcpy;
-#endif
+    convertor->cbmemcpy = &opal_convertor_accelerator_memcpy;
 #if OPAL_ROCM_SUPPORT
```
clean up the rocm bits while you're here.

bwbarrett commented on opal/datatype/opal_datatype_copy.c:
```diff
@@ -67,78 +55,104 @@ static size_t opal_datatype_memop_block_size = 128 * 1024;
+    }     
+}
+
+static void *opal_datatype_accelerator_memmove(void *dest, const void *src, size_t size)
```
Why does this function make sense over just blindly calling memmove with type UNSPECIFIED?

wckzhang replied:
Because the NULL component doesn't do memcpy or memmove. Your ask was for check_addr to return 0 always, and I figured if we were going to do the check anyway, it would be more efficient to preserve that data and pass it as a type.

bwbarrett replied:
Looking at all the uses, this function included, I may have been wrong.  It's weird to me that we're pushing complexity to the upper layers (read: many places) rather than handling it in the accelerator framework.  This is definitely one of the places that feels like we made the wrong decision about the complexity of the two implementation options.

wckzhang replied:
Changed the null component to do the memcpy and memmoves

bwbarrett reviewed (changes requested):
There are three serious problems with this patch series.  First, ROCM support is now in OMPI, so we need a ROCM component and the patches that remove CUDA code should also remove ROCM code.  We can't have a half and half situation.  Discuss with @edgargabriel how to make that happen.

Second, I don't believe that this patch series is bisectable from a runtime perspective.  You have patches that change the datatype engine but the rest of the code still uses the old CUDA code.  I think you're going to have to have fewer, larger patches.  That's unfortunate, but I think the reality we face.

Third, you don't actually remove the old CUDA code.  It all needs to die as part of this patch series.

bwbarrett commented:
I forgot to add that you have a PMIx submodule change in the PR for no apparent reason; that's almost certainly a mistake.

bwbarrett commented:
The failures look real; the datatype unit tests are failing with a segmentation fault.

a-szegel commented:
Failures are real, first failing commit: [pml/ob1: Replace cuda code with accelerator code](https://github.com/open-mpi/ompi/pull/10069/commits/d90184fecdf9f0e2a05f0a78ee55ef14b526e1b0)

a-szegel commented on opal/mca/accelerator/base/help-accelerator-base.txt:
```diff
@@ -0,0 +1,14 @@
+# $HEADER$
+#
+[Multiple Accelerators Found]
+ERROR: Both %s and %s accelerator frameworks were successfully initialized,
```
There can only be two that initialize successfully, the NULL component and one user defined component.  I can change this to say 1 non-null component if that reads better (or completely ignore the NULL component, and say "only 1 type of accelerator is supported")

bwbarrett replied:
This error message is explicitly for when more than 1 (non-NULL) component is found (for example, if there are AMD and Nvidia cards in a single box, with no mca parameters).  I don't see why you couldn't have a box with AMD, Nvidia, and Ze cards in it?

edgargabriel replied:
Consumer level cards? Maybe. For the more high-end cards it will probably not make sense however, since you will not be able to take advantage of the GPU interconnect (e.g. Nvlink, XGMI) between cards of different vendors, and they are a big reason for why you can scale to multiple GPUs per node for many jobs.

a-szegel replied:
Updated selection logic to make it simple for the user to figure out their bug.  New help text:
```
[Multiple Accelerators Found]
ERROR: Open MPI only supports one accelerator framework per node. Please pick one of the
initialized accelerator components via command line argument "--mca accelerator <component>".
The following accelerators initialized successfully: %s.
The program will abort to allow you to resolve the accelerator conflict.
```

a-szegel commented on opal/mca/accelerator/accelerator.h:
```diff
@@ -0,0 +1,480 @@
+ *
+ * @return                   OPAL_SUCCESS or error status on failure
+ */
+typedef int (*opal_accelerator_base_module_vector_memcpy_fn_t)(
```
supports both

a-szegel replied:
Switching name to matrix.

wckzhang replied:
Moving matrix memcpy to a separate PR, george has ACK'd this.

a-szegel commented on opal/mca/accelerator/base/base.h:
```diff
@@ -0,0 +1,38 @@
+#define MCA_ACCELERATOR_BASE_H
+
+#include "opal_config.h"
+#include "opal/types.h"
```
removed types.h

a-szegel commented on opal/mca/accelerator/accelerator.h:
```diff
@@ -0,0 +1,480 @@
+ *
+ * @param[IN] addr           Pointer to check
+ * @param[OUT] dev_id        Returns the device id against which the memory was allocated
+ * @param[OUT] flags         May be set to indicate additional information
```
I think the design was that the user initializes to a known empty state of 0, and if any bits are set to 1, a flag has been set.  I think you are probably probing for better documentation here vs a change in functionality.  Will update the documentation to reflect how it is actually used.

a-szegel replied:
Updated to:
```
 * @param[OUT] flags         NOTE: This functionality is specific to component implementation.
 *                                            The user is required to pass in a uint64_t initialized to 0.
 *                                            Component implementation can optionally set flags to indicate
 *                                            additional information about the corresponding pointer. such
 *                                            as whether it belongs to unified memory.
 ```

wckzhang replied:
The flags are defined in accelerator.h @tschuett. Only one flag is defined but it's in this file: ```#define MCA_ACCELERATOR_FLAGS_UNIFIED_MEMORY 0x00000001```

a-szegel replied:
Updated description and impl to match discussion.

a-szegel commented on opal/mca/accelerator/accelerator.h:
```diff
@@ -0,0 +1,480 @@
+ * Allocates size bytes memory from the device and sets ptr to the
+ * pointer of the allocated memory. The memory is not initialized.
+ * If size is 0, then the function will return 0 and ptr will not be set.
+ * If multiple devices are present, a component can decide which
```
Updated comment to reflect new behavior in PR: https://github.com/wckzhang/ompi/pull/7

a-szegel replied:
I think the reason why this doesn't take a device ID is b/c CUDA implementation assumes that the device ID is present in the context that is already set, i.e.. you can only malloc in your current context.

a-szegel replied:
The CUDA component would ignore dev_id which could be confusing but I think it still makes sense to have here.

wckzhang replied:
I'm thinking about abstracting away all the dev_id stuff and let the component impl figure it out. I think the main concern is if ZE can handle this properly, as ROCm and cuda can.

wckzhang replied:
I added dev_ids back in for more flexibility.

a-szegel commented on opal/mca/accelerator/base/accelerator_base_frame.c:
```diff
@@ -0,0 +1,52 @@
+ */
+
+#include "opal_config.h"
+#include "opal/constants.h"
```
Corrected spacing in all files.

a-szegel replied:
Only fixed things in the accelerator framework, realized there are still issues in other sections of the code!

a-szegel commented on opal/mca/accelerator/base/accelerator_base_frame.c:
```diff
@@ -0,0 +1,52 @@
+
+#include "opal/mca/base/base.h"
+#include "opal/mca/mca.h"
+#include "opal/util/output.h"
```
removed.

a-szegel commented on opal/mca/accelerator/base/accelerator_base_select.c:
```diff
@@ -0,0 +1,138 @@
+#include "opal/mca/mca.h"
+#include "opal/mca/accelerator/base/base.h"
+#include "opal/mca/accelerator/accelerator.h"
+#include <string.h>
```
fixed everywhere.

a-szegel commented on opal/mca/accelerator/null/accelerator_null_component.c:
```diff
@@ -0,0 +1,307 @@
+
+static size_t accelerator_null_get_handle_size(void)
+{
+    return OPAL_ERR_NOT_IMPLEMENTED;
```
Updated to be 0

a-szegel commented on opal/mca/accelerator/null/accelerator_null_component.c:
```diff
@@ -0,0 +1,307 @@
+
+static bool accelerator_null_is_ipc_enabled(void)
+{
+    return OPAL_ERR_NOT_IMPLEMENTED;
```
updated to return false.

a-szegel commented on opal/mca/accelerator/cuda/accelerator_cuda.c:
```diff
@@ -0,0 +1,690 @@
+#include "opal/mca/accelerator/base/base.h"
+#include "opal/util/show_help.h"
+#include "opal/util/proc.h"
+#include <cuda.h>
```
Updated Headers globally.

a-szegel commented on opal/mca/accelerator/cuda/owner.txt:
```diff
@@ -0,0 +1,7 @@
+# owner: institution that is responsible for this package
+# status: e.g. active, maintenance, unmaintained
+#
+owner: AMAZON
```
Removed AMAZON ownership :)

bwbarrett replied:
I still see this as owned by Amazon in the diff...

wckzhang replied:
Huh, changed

wckzhang commented on opal/mca/accelerator/accelerator.h:
```diff
@@ -0,0 +1,480 @@
+    opal_accelerator_stream_t *stream);
+
+/**
+ * Creates an event. You must allocate memory for event prior to
```
Streams are sequential executing command queues where we place memcpy's asynchronously. Events are events that we can append to a stream after a memcpy command so we can query or synchronize the event (and thus query/synchronize the asynchronous memcpy).

wckzhang replied:
There is at the top of the file which gives an explanation of Transfer types, Streams and Events, Asynchronous copies, and host register/unregister

wckzhang commented on opal/mca/accelerator/accelerator.h:
```diff
@@ -0,0 +1,480 @@
+ *
+ * @return                   OPAL_SUCCESS or error status on failure.
+ */
+typedef int (*opal_accelerator_base_module_destroy_event_fn_t)(
```
For component specific destructor functions

a-szegel commented:
Initial Comment Cleanup: https://github.com/wckzhang/ompi/pull/9

wckzhang commented on opal/mca/accelerator/accelerator.h:
```diff
@@ -0,0 +1,480 @@
+ *
+ * @return                   OPAL_SUCCESS or error status on failure.
+ */
+typedef int (*opal_accelerator_base_module_record_event_fn_t)(
```
It writes an event to a stream, I could probably have a better explanation for all of this somewhere.

wckzhang replied:
Will add a note

qkoziol replied:
The description here is adequate for the common usage for accelerators  (e.g. docs for CUDA Graph events is similar: https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#events)

tschuett commented:
Would it be a win to have a accelerator.rst explaining the design, i.e., streams and events.

wckzhang commented:
> The failures look real; the datatype unit tests are failing with a segmentation fault.

Yeah, I think I messed something up in my pml ob1 code conversion, I'll fix that

wckzhang commented:
> Would it be a win to have a accelerator.rst explaining the design, i.e., streams and events.
> 
> Sorry, I meant to say: would an introduction in accelerator.h help?

I'll add some more help texts

a-szegel commented on opal/mca/accelerator/cuda/accelerator_cuda.c:
```diff
@@ -0,0 +1,690 @@
+
+static bool accelerator_cuda_is_ipc_enabled(void)
+{
+    return true;
```
How do we handle instances with one accelerator?

wckzhang replied:
I was thinking that some API's wouldn't have IPC support so this could be checked. I suppose if it's really necessary it can be added when someone doesn't support IPC

wckzhang replied:
Going to move IPC code to a separate PR

a-szegel commented on opal/runtime/opal_init.c:
```diff
@@ -42,6 +42,7 @@
 #include "opal/include/opal_config.h"
 
 #include "opal/datatype/opal_datatype.h"
+#include "opal/mca/accelerator/base/base.h"
```
We can't do this actually.  The selection logic relies on the NULL component existing, and the NULL component comes in the second commit.  Either we re-work the selection logic, or deal with the dead code on the first commit.

bwbarrett replied:
Then add the NULL component in this commit?

a-szegel commented on opal/mca/accelerator/cuda/accelerator_cuda.h:
```diff
@@ -0,0 +1,69 @@
+} opal_accelerator_cuda_component_t;
+
+/* Structure to hold CUDA function pointers that get dynamically loaded. */
+struct  accelerator_cuda_func_table {
```
Does anyone have any objections to just dlopening libcuda and getting rid of the function table (since it will now be guarded behind the CUDA component of the accelerator framework) 

@janjust

janjust replied:
@open-mpi/cuda

a-szegel replied:
So after doing more research into this, we still need to keep the CUDA function table.  The reason is that we want to support the package manager use case.  The package managers want a openMPI library compiled with --with-cuda, and they want to be able to run it on machines that don't have CUDA libraries.  If we want that use case to work, we cannot link cuda into our application at all (b/c then the linker would fail to find the library, and the program wouldn't start).

The way dlopen, dlsym, and dlclose work for loading dynamic libraries at runtime is they return the symbols address, i.e..:
```
handle = dlopen(“foo.so”, RTLD_NOW | RTLD_LOCAL);
(void (*)(void)) foo = (void (*)(void)) dlsym(handle,”the_foo_function”);
```

If someone wants to call the the_foo_function, they must have access to the function pointer returned from dlsym (i.e.. the function table).

bwbarrett replied:
That seems motivation to have the cuda component built as a dso instead as part of libopen-pal, and you get the same thing.

bwbarrett replied:
I'm going to state this stronger.  I don't see any reason for the function table.  There's no reason that cuda should be different than OFI or UCX or any of the other multitude of components with external requirements.

tschuett commented:
Can I query the accelerator framework for the number of devices of the current process? Or is the assumption that it is either zero or 1? Or do I have to query the underlying component?

a-szegel commented:
The accelerator framework doesn't quite isolate the application from the accelerator hardware.  For CUDA, user is responsible for setting up the CUDA context before initializing the framework (which will tell the framework which accelerator to use).  It would be best if we could fix this breach in encapsulation but we would have to modify the MPI api's which is not an option right now.

shijin-aws commented on opal/mca/accelerator/accelerator.h:
```diff
@@ -0,0 +1,475 @@
+ *
+ * @return                   OPAL_SUCCESS or error status on failure.
+ */
+typedef int (*opal_accelerator_base_module_create_event_fn_t)(
```
For habana you also need `device_id` when you create an event.

```
synStatus SYN_API_CALL synEventCreate( synEventHandle*       pEventHandler,
                                       const synDeviceId     deviceId,
                                       const uint32_t        flags );
```

bwbarrett replied:
The API needs to make a decision and stick with it.  Either there's one device associated with a process and none of the calls require a device id (other than maybe a get/set_device_id) or they all need a device id.  I don't have strong preferences, but the argument that the underlying API has a device id doesn't make sense; instead it should be an argument about the framework's usage model.

wckzhang replied:
We decided to avoid calls with device_ids. I removed device_id from most of the API's other than the get device id. If a new component comes along with a different requirement of device_ids that cannot be adapted to the API comes along, then they can propose some changes and we can have a v2 of the API.

shijin-aws commented on opal/mca/accelerator/accelerator.h:
```diff
@@ -0,0 +1,475 @@
+ *
+ * @return                   OPAL_SUCCESS or error status on failure
+ */
+typedef int (*opal_accelerator_base_module_matrix_memcpy_fn_t)(
```
which accelerator type need such API?

wckzhang replied:
cuda has a 2d memcpy, it's for optimization purposes for something like collectives

qkoziol replied:
Is the matrix_memcpy function pointer actually used within OpenMPI anywhere?   This looks like something that is unnecessary and should be removed.  CUDA also has a 3D memcpy, but adding CUDA-specific special things to the function pointers isn't a great idea, unless they are actually used in this PR.   (As mentioned earlier, if desired, they can be added in a future 'v2' of this internal API)

wckzhang replied:
It's not currently used. This was a request from @bosilca to improve collectives. It shouldn't be too much of a problem to keep it in since a matrix memcpy can be implemented from regular memcpy afaik

wckzhang replied:
Moved matrix memcpy out of this PR.

shijin-aws commented on opal/mca/accelerator/accelerator.h:
```diff
@@ -0,0 +1,475 @@
+ *
+ * @return                   OPAL_SUCCESS or error status on failure
+ */
+typedef int (*opal_accelerator_base_module_free_fn_t)(
```
habana needs a device id to free device buffer.

shijin-aws replied:
Unfortunately there is not such an API in habana yet

Habana does not have an exposed API to query which device is being used in current process, or which device is on for a given device buffer.

Application needs to acquire a device with device id explicitly, and pass the device id to open mpi.

a-szegel replied:
We cannot touch the MPI application API because it takes years to get changes in (with the current MPI API, it is impossible for the user to pass dev_id through the interface).  Under the current design constraints, the Habana accelerator component would need some global hack (similar to a CUDA context) so that it can fetch the dev_id it needs.

It is unfortunate that this "hack"/"abstraction failure" ties Applications to specific accelerator hardware.

This hack could be as simple as:
```
/* Habana app.c */
int habana_dev_id = 0;

/* Habana Accelerator Component */
extern int habana_dev_id;
```

bwbarrett replied:
I'd strongly recommend environment variables (or, better yet, MCA parameters!) for this, not an extern.  But, again, uplevel the conversation to the framework's expected use behaviors.

edgargabriel replied:
how about an info object (MPI_Comm_set_info, MPI_Win_set_info/Win_create)?

tschuett replied:
If you remove the device id from the API and every component has to handle it themselves, then your invariant is that there is exactly one device per process.

wckzhang replied:
Just as a note, we're not actually thinking about adding Habana support right now, I was just trying to see how generic our API was, since the current cuda implementation doesn't rely on device ids. I'm starting to lean on the boat of getting rid for device ids entirely and letting the implementations handle it, but setting info objects could be a solution.

shijin-aws commented on opal/mca/accelerator/accelerator.h:
```diff
@@ -0,0 +1,475 @@
+ * @return                   OPAL_SUCCESS or error status on failure
+ *
+ */
+typedef int (*opal_accelerator_base_module_get_handle_fn_t)(
```
I was confused by what handle it is by reading the libfabric's implementation. If it is for IPC only, can we rename it as `*get_ipc_handle`? If it's not for IPC only, what are the other user cases for this function?

wckzhang replied:
I can do that

wckzhang replied:
Moved IPC out of this PR.

shijin-aws commented on opal/mca/accelerator/accelerator.h:
```diff
@@ -0,0 +1,475 @@
+ *
+ * @return                   OPAL_SUCCESS or error status on failure
+ */
+typedef int (*opal_accelerator_base_module_host_register_fn_t)(
```
for habana, you need to pass device id to register a host buffer:
```
synStatus SYN_API_CALL synHostMap( const synDeviceId          deviceId,
                                   const uint64_t             size,
                                   const void*                buffer );
```

wckzhang replied:
Added device ID to each of the API's including host register.

shijin-aws commented on opal/mca/accelerator/accelerator.h:
```diff
@@ -0,0 +1,475 @@
+ * @return                   OPAL_SUCCESS or error status on failure
+ */
+typedef int (*opal_accelerator_base_module_host_unregister_fn_t)(
+    void *ptr);
```
Similar comments as the register one:

```
synStatus SYN_API_CALL synHostUnmap( const synDeviceId    deviceId,
                                     const void*          buffer );
```

shijin-aws commented on opal/mca/accelerator/accelerator.h:
```diff
@@ -0,0 +1,475 @@
+ *
+ * @return                   OPAL_SUCCESS or error status on failure
+ */
+typedef int (*opal_accelerator_base_module_create_stream_fn_t)(
```
For habana, you also need to specify the stream type for different directions (h-d, d-h, d-d, h-h)

```
synStatus SYN_API_CALL synStreamCreate( synStreamHandle*       pStreamHandle,
                                        const synDeviceId      deviceId,
                                        const synStreamType    streamType,
                                        const uint32_t         flags );

```

wckzhang commented:
TODOs:

1. rm -rf opal/cuda -> move all usages of cuda in OMPI to Accelerator Framework
2. Figure out how to handle multiple Accelerators (of same type) on an instance, does dev_id belong in the interface? 
~~3. Fix Converter interface instead of messing with the converter flags <-- DONE~~
~~4. Add  check capabilities interface to accelerator: bool check_capabilities( void *ptr, size_t size, uint32_t flags) + remove check_addr calls <-- Abandoned~~
5. Figure out why accelerator cleanup code is never being executed
6. Make Commits bisectable from a runtime perspective, i.e.. fewer larger commits
7. Fix OFI MTL's configure.m4
8.  ompi/mca/mtl/ofi/mtl_ofi_component.c: printf is now going to always fire for providers that don't support FI_HMEM.
~~9. Add documentation about event / stream behavior to interface <-- DONE~~
~~10. opal/mca/accelerator/base/help-accelerator-base.txt: Output a list of every component that failed to be created <-- DONE~~
11. Remove accelerator_cuda_func_table
12. Add host memcpy & memmove functionality to NULL component + remove check_addr calls
~~13. Change create_stream to not allocate <-- DONE~~
14. Decide if accelerator_cuda_is_ipc_enabled() stays in the interface
15. Add ROCM Component (Implemented by Edgar) <-- Edgar is doing this
16. rm -rf opal/ROCm (Implemented by Edgar) <-- Edgar is doing this
~~17. Rename handle functions ipc_handle (unclear) <-- DONE~~

ibm-ompi commented:
The IBM CI (GNU/Scale) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/3677ad9fc8679de2828bd1432ecf7eb8

ibm-ompi commented:
The IBM CI (XL) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/3434b52b1662802e492cca83ade266b2

Akshay-Venkatesh commented on ompi/mca/pml/ob1/pml_ob1_accelerator.c:
```diff
@@ -0,0 +1,599 @@
+
+    OBJ_DESTRUCT(&pml_ob1_accelerator_htod_lock);
+    OBJ_DESTRUCT(&pml_ob1_accelerator_dtoh_lock);
+}
```
@wckzhang Most the of the above functions and structures have htod and dtoh variants. Code duplication can be reduced by introducing htod, dtoh enums and passing that as argument to functions or having that as loop index. What do you think?

wckzhang replied:
Yeah it can, I'm trying to (it's not working too well) make as few changes as possible. This PR is big as is, and I'd like it to get merged before we do too many changes.

wckzhang replied:
In the interest in getting this PR moved, I'm not going to make these changes.

Akshay-Venkatesh commented on ompi/mca/pml/ob1/pml_ob1_accelerator.c:
```diff
@@ -0,0 +1,599 @@
+                                               mca_bml_base_btl_t* bml_btl,
+                                               size_t size) {
+    int rc;
+#if OPAL_CUDA_GDR_SUPPORT
```
This ifdef region is not CUDA-specific right? It should probably apply to all accelerator instances.

wckzhang replied:
I think this path would need to be tested with Rocm, I'm not exactly sure if this is safe to take out so I left it in

wckzhang replied:
As follow up work, will need to go test this with other accelerators, leaving for now.

Akshay-Venkatesh commented on ompi/mca/pml/ob1/pml_ob1_cuda.c:
```diff
@@ -1,222 +0,0 @@
-            ep->btl_send.bml_btls[i].btl_flags |= MCA_BTL_FLAGS_CUDA_GET;
-        }
-    }
-}
```
@wckzhang After removing this code, have you tested if PML ob1 works as before with CUDA?

wckzhang replied:
This code wasn't really removed, just moved into pml_ob1_accelerator.c and renamed.

Akshay-Venkatesh commented on opal/datatype/opal_datatype_copy.c:
```diff
@@ -157,19 +117,16 @@ int32_t opal_datatype_copy_content_same_ddt(const opal_datatype_t *datatype, int
 
-    fct = non_overlap_copy_content_same_ddt;
-    SET_GPU_COPY_FCT(device_bufs, fct, non_overlap_gpu_copy_function);
+    fct = non_overlap_accelerator_copy_content_same_ddt;
```
does this function handle non-device buffers as well now? Previously we conditionally set copy function to accelerator type.

wckzhang replied:
Yes

Akshay-Venkatesh reviewed (commented):
Hi @wckzhang. From NVIDIA side, we have some comments on the overall implications of the new framework - 

Basic questions (for us to test this framework):
- What are the configure parameters to build OMPI with accelerator component?
- What are the runtime parameters needed to toggle between accelerator instances?


Framework questions:
- Can you confirm that smcuda will continue to work with accelerator component?
- In the absence of smcuda, can cuda-MPI work with just the accelerator framework? If yes, which btl/mtl/pml combinations work?


Some thoughts on this component's extensions that doesn't need to be part of this PR but the component should allow for in the future:
- Does the framework allow for multiple GPUs to be selected by a single MPI rank? I didn't see device IDs being mentioned in accelerator component function signatures or data structures.
- Does the framework make assumptions that would not allow for stream extensions of MPI communication routines? There are some static streams setup for the purposes of htod/dtoh transfers for internal MPI use, but there should be provisioning for use of user provided streams/events.

About merging the PR for OMPI 5.0 from NVIDIA's testing viewpoint:
- The PR hasn't been tested with ob1 + smcuda which is still used users owing to its relative stability and as a point of comparison with other options
- The PR hasn't been tested with PML UCX which is the most used component for cuda-MPI workloads.

We can commit to testing the above cases and provide feedback but we would need time to rigorously test large changes proposed by this PR.

wckzhang commented:
> * What are the configure parameters to build OMPI with accelerator component?
--with-cuda=<path_to_cuda>

> * What are the runtime parameters needed to toggle between accelerator instances?
Not sure if you mean between different components or if you mean between different devices.
There's no runtime switching of devices from mpi point of view (for now) Cuda contexts will be set by the application (MPI has bad communication with applications about GPU information) and Open MPI will perform tasks under the current cuda context

> Framework questions:
> 
> * Can you confirm that smcuda will continue to work with accelerator component?
I will have to do some changes but it should work after I do those. 

> * In the absence of smcuda, can cuda-MPI work with just the accelerator framework? If yes, which btl/mtl/pml combinations work?
The (ofi mtl and pml cm) has been tested by me and works. Other than that I can test the tcp btl which would not work. One thing to consider is that smcuda was branched off of the old sm component, which performs a lot worse than the new sm component (formerly known as vader). For long term smcuda support, should consider rewriting it to be integrated with the new sm component, or at least branched off of the new component.
 
> Some thoughts on this component's extensions that doesn't need to be part of this PR but the component should allow for in the future:
> 
> * Does the framework allow for multiple GPUs to be selected by a single MPI rank? I didn't see device IDs being mentioned in accelerator component function signatures or data structures.
No, with the way cuda contexts are set by the application (and rocm functions similarly), switching devices didn't really make sense as MPI doesn't have a lot of information other than the current cuda context. As far as I could tell, most cuda things didn't need a specified device id. I had formerly had device id in the API, but neither rocm nor cuda seemed to require it, and without a working implementation, it was hard for me to imagine which API's needed extension. When the need arises, we can add device_ids into the API.

> * Does the framework make assumptions that would not allow for stream extensions of MPI communication routines? There are some static streams setup for the purposes of htod/dtoh transfers for internal MPI use, but there should be provisioning for use of user provided streams/events.

There are functions for creating streams and events in the accelerator API. That's only for MPI internals though. Unfortunately MPI doesn't allow for much communication to the user about streams/events. We'd have to add some info keys but that would make it very Open MPI specific. There's a community that is trying to push for better information exchange in the MPI forum.

> About merging the PR for OMPI 5.0 from NVIDIA's testing viewpoint:

There isn't currently a plan to merge the PR into OMPI 5.0, unless 5.0 is extremely delayed (which is not in our best interests anyway).

> * The PR hasn't been tested with ob1 + smcuda which is still used users owing to its relative stability and as a point of comparison with other options

My next update to this PR will update smcuda and will properly tested. I can do some testing with tcp + btl smcuda configurations.

> * The PR hasn't been tested with PML UCX which is the most used component for cuda-MPI workloads.

I haven't and probably won't be able to test with the PML UCX. I'm not too familiar with how the UCX PML cuda functionalities. I looked at ompi_pml_ucx.cuda_initialized but it seemed unused. I assume most of the cuda stuff happens under the covers in UCX itself, if you know more, can you elaborate?

wckzhang commented on ompi/mca/pml/ucx/pml_ucx.c:
```diff
@@ -265,26 +262,14 @@ int mca_pml_ucx_open(void)
 
-#if HAVE_UCP_ATTR_MEMORY_TYPES && OPAL_CUDA_SUPPORT
-    if (attr.memory_types & UCS_BIT(UCS_MEMORY_TYPE_CUDA)) {
-        mca_common_cuda_stage_one_init();
```
Yes, it has dl_open logic during the component init phase, see https://github.com/open-mpi/ompi/pull/10069/files#diff-fb4497b2fa6fedf5c529388bfd346a5561f3c6555319f14f3d53aca44ab31f0fR298

It doesn't need to be called multiple times because the init is just called during opal init. I haven't tested with PML UCX, but iirc PML UCX didn't actually need common_cuda, but maybe someone who worked on pml ucx could elaborate on that better.

wckzhang replied:
Yes for testing, Akshay helped ttest that path. Keeping with dl_open right now but looking into linking libcuda.so.1 directly.

wckzhang commented on ompi/mca/pml/ob1/pml_ob1_recvreq.c:
```diff
@@ -309,9 +304,11 @@ static int mca_pml_ob1_recv_request_ack(
-           rdma_num != 0) {
+        if (opal_convertor_need_buffers(&recvreq->req_recv.req_base.req_convertor) == 0 &&
+            !(recvreq->req_recv.req_base.req_convertor.flags & CONVERTOR_ACCELERATOR) &&
+            !(recvreq->req_recv.req_base.req_convertor.flags & CONVERTOR_ACCELERATOR_UNIFIED) &&
```
I can do that

wckzhang replied:
Added the comment.

ibm-ompi commented:
The IBM CI (XL) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/ibm-ompi/88b480fb98dfd2bdac939f023e1f8fe6

awlauria commented:
It looks like a failure in `make check` . Judging by the tests crashing, I would guess the issue in in the opal datatype engine.

wckzhang commented:
The issue is that the datatype unit tests only do opal_init_util and opal_datatype_init while the accelerator is initialized in the opal_init function. Since there's a dependency in the opal_convertor on the accelerator being initialized, it segfaults. Theoretically moving the accelerator framework initialization into opal_init_util should solve this issue.

wckzhang commented:
The Mellanox CI seems to be an infrastructure issue, can someone help me retrigger it?

janjust commented:
/azprun

wckzhang commented:
Is the mellanox CI node down?

janjust commented:
@artemry-nv Can you please check, is the Mellanox CI down?

artemry-nv commented:
> @artemry-nv Can you please check, is the Mellanox CI down?

There's an issue with our lab infrastructure - IT is working on it. I'll inform when it's fixed.

artemry-nv commented:
> @artemry-nv Can you please check, is the Mellanox CI down?

Should work fine now.
Sorry for the inconvenience.

jjhursey commented:
The instructions for retesting are [on the wiki](https://github.com/open-mpi/ompi/wiki/PRJenkins#triggering-a-retesting-by-all-or-a-subset-of-ci).

We see this PRRTE error periodically in MTT. My best guess is a race condition in the startup sequence, but we haven't had the cycles to look into it.

bot:ibm:retest

wckzhang commented:
@Akshay-Venkatesh Do you have any updates on the testing front?

wckzhang commented:
Rebased, testing was approved from nvidia. Does anyone object with me merging and letting it bake in main for a bit after CI passes?

edgargabriel reviewed (approved):
I ran the branch through our testsuite, I do see two tests failing that didn't use to fail in my earlier tests. But I think that its probably more beneficial to commit the branch and work on the issues afterwards, instead of further delaying merging the branch to main. Furthermore, it is unclear whether its something that has changed in rocm, ompi or ucx that causes those tests to fail.

wckzhang commented:
@bwbarrett I think merging is blocked on requested changes

bwbarrett reviewed (changes requested):
It looks like a number of my comments remain unaddressed (around objects vs. structure and around device ids).  These need to be resolved before committing to an API.

wckzhang commented on opal/mca/accelerator/accelerator.h:
```diff
@@ -0,0 +1,510 @@
+    void *stream;
+};
+typedef struct opal_accelerator_stream_t opal_accelerator_stream_t;
+OBJ_CLASS_DECLARATION(opal_accelerator_stream_t);
```
Okay, I will take a look at this

bwbarrett commented on opal/mca/accelerator/accelerator.h:
```diff
@@ -0,0 +1,515 @@
+ *
+ */
+typedef int (*opal_accelerator_base_module_get_ipc_handle_fn_t)(
+    void *dev_ptr, void **handle);
```
The handles should have types (even if they're typedefs to void*).  Who frees the memory and how (ie, document the lifespan of the returned handle as part of this function documentaiton).

wckzhang replied:
As ipc has been moved out of this PR, not relevant.

bwbarrett reviewed (changes requested):
I think you've actually gone backwards with the interface.  You had the right approach with OPAL objects for the objects exposed by the interface, but removed all that for some reason.

Also, I feel like you've not done a good job of addressing the other comments made in the previous feedback and following up to close the loop.

wckzhang commented:
I'll try to address the rest of Brian's comments this week

qkoziol commented on opal/mca/accelerator/base/owner.txt:
```diff
@@ -0,0 +1,7 @@
+# owner: institution that is responsible for this package
+# status: e.g. active, maintenance, unmaintained
+#
+owner:
```
Should Amazon have an owner here?

qkoziol commented on opal/mca/accelerator/null/accelerator_null_component.c:
```diff
@@ -0,0 +1,306 @@
+/* Accelerator API's Implementation */
+static int accelerator_null_check_addr(const void *addr, uint64_t *flags)
+{
+    /* Always return that the pointer belongs to the host */
```
Shouldn't this set *flags to '0' then?   (return 0 is about errors)

qkoziol replied:
Oops, I'm wrong here.

qkoziol reviewed (changes requested):
Agree with Brian about getting rid of the function pointer table

qkoziol commented on opal/datatype/opal_convertor.c:
```diff
@@ -560,19 +551,34 @@ size_t opal_convertor_compute_remote_size(opal_convertor_t *pConvertor)
     }
-#elif OPAL_ROCM_SUPPORT
-    if (!(convertor->flags & CONVERTOR_SKIP_GPU_INIT)) {
+#if OPAL_ROCM_SUPPORT
```
Delete?

qkoziol commented on opal/datatype/opal_convertor.c:
```diff
@@ -613,12 +619,11 @@ int32_t opal_convertor_prepare_for_send(opal_convertor_t *convertor,
     }
-#elif OPAL_ROCM_SUPPORT
-    if (!(convertor->flags & CONVERTOR_SKIP_GPU_INIT)) {
+#if OPAL_ROCM_SUPPORT
```
Delete?

qkoziol commented on opal/datatype/opal_convertor.c:
```diff
@@ -742,10 +747,10 @@ void opal_convertor_dump(opal_convertor_t *convertor)
     }
-    if (convertor->flags & CONVERTOR_CUDA) {
+    if (convertor->flags & CONVERTOR_ACCELERATOR) {
         opal_output(0, "CUDA ");
```
"CUDA"  => "Accelerator"

qkoziol replied:
And below

qkoziol commented on opal/datatype/opal_convertor.h:
```diff
@@ -183,10 +181,6 @@ static inline int32_t opal_convertor_need_buffers(const opal_convertor_t *pConve
 {
     if (OPAL_UNLIKELY(0 == (pConvertor->flags & CONVERTOR_HOMOGENEOUS)))
         return 1;
-#if OPAL_CUDA_SUPPORT || OPAL_ROCM_SUPPORT
```
Should this be retained, with a change to checking the CONVERTOR_ACCELERATOR & CONVERTOR_ACCELERATOR_UNIFIED flags?

wckzhang replied:
We moved this functionality upward outside this function.

qkoziol reviewed (approved):
Small changes needed, generally OK though

qkoziol commented on opal/mca/accelerator/cuda/accelerator_cuda_component.c:
```diff
@@ -0,0 +1,333 @@
+        return NULL;
+    }
+
+    if (OPAL_SUCCESS != accelerator_cuda_populate_func_table(libcuda_handle)) {
```
I noticed that the ROCM init routine acquires a lock around a similar set of code in its init routine.   Is one needed here?   (Should the lock in the ROCM init routine be removed @edgargabriel ?)

bwbarrett replied:
The accelerator init functions will all be during opal_init, which is serialized.  So I don't think a lock is needed here in this case.  It would be needed in the lazy init case that used to exist when this code was in opal/cuda.

wckzhang replied:
I got rid of the locking in the init routine

qkoziol reviewed (changes requested):
Agree with Brian about getting rid of the function pointer table

qkoziol commented on opal/mca/accelerator/rocm/accelerator_rocm_module.c:
```diff
@@ -0,0 +1,623 @@
+
+    stream->stream = (hipStream_t *)malloc(sizeof(hipStream_t));
+    if (NULL == stream->stream) {
+        free(stream);
```
I don't think that the stream should be freed here (anymore?)

qkoziol replied:
And below

wckzhang replied:
This was a bug, but I changed the API as well so fixed these alongside.

qkoziol commented on opal/mca/accelerator/rocm/accelerator_rocm_module.c:
```diff
@@ -0,0 +1,623 @@
+
+    event->event = malloc(sizeof(hipEvent_t));
+    if (NULL == event->event) {
+        free(event);
```
I don't think the event should be freed here.

qkoziol commented on opal/mca/accelerator/cuda/accelerator_cuda.c:
```diff
@@ -0,0 +1,999 @@
+
+    event->event = malloc(sizeof(CUevent));
+    if (NULL == event->event) {
+        free(event);
```
I don't think the event should be freed here.

wckzhang replied:
Yeah this is a bug due to an API change

qkoziol reviewed (changes requested):
Agree with Brian about getting rid of the function pointer table

qkoziol commented on opal/mca/accelerator/rocm/accelerator_rocm_module.c:
```diff
@@ -0,0 +1,623 @@
+        return OPAL_ERR_BAD_PARAM;
+    }
+
+    hipError_t err = HIP_FUNCS.hipEventRecord((hipEvent_t)event->event,
```
I believe that the first parameter should be *(hipEvent_t *)event->event here.

wckzhang replied:
@edgargabriel is this still applicable?

edgargabriel replied:
yes, I missed that one. Shall I make the patch or would you like to do that?

qkoziol reviewed (changes requested):
Some small bugs, but generally seems in good shape

wckzhang commented:
In the interest of making progress, I removed some API's to put into separate PR's and also am going to re-add common cuda and do the libcuda linking directly later. As this PR has gotten too large, it's hard to merge without additional comments, I will follow up with those PR's as separate pieces.

wckzhang commented:
The Mellanox CI failure looks real, I'm guessing it has to do with some of the changes from opal util/core that had some conflicts when I rebased

wckzhang commented:
Yeah the CI failure is the same as I hit before. Basically the datatype unit tests only call opal_init_util and ompi_datatype_init (which calls opal_datatype_init). The datatype has a dependency on the accelerator framework, but it's not initialized in either. I moved the accelerator framework opening to opal_datatype_init to fix.

wckzhang commented:
Or it may be more appropriate for the unit tests to call opal_init and not just opal_init_util

wckzhang commented:
/azp run

edgargabriel reviewed (approved):
double checked mostly the rocm component and ompio stuff, both look good as far as I can tell.

wckzhang commented:
Mellanox CI is having a hard time with docker pull, should I keep retrying the pipeline or wait until a fix comes in?

jsquyres commented:
@wckzhang Wait until a fix comes in.  NVIDIA is aware of the problem; @janjust is working on it.  This is affecting all OMPI PRs, not just this PR.

bwbarrett commented on opal/mca/accelerator/cuda/configure.m4:
```diff
@@ -0,0 +1,33 @@
+          [$2])
+
+])dnl
+
```
unnecessary trailing newline

wckzhang replied:
Fixed

bwbarrett reviewed (approved):
You have a couple open conversations you should resolve (none are conversations I started, I think).  This otherwise looks good to me.

wckzhang commented on opal/mca/accelerator/accelerator.h:
```diff
@@ -0,0 +1,480 @@
+ *
+ * @return                   OPAL_SUCCESS or error status on failure
+ */
+typedef int (*opal_accelerator_base_module_destroy_stream_fn_t)(
```
Converted them to true objects with destructor functions and adjusted the API to remove the destroy functions.

wckzhang commented:
One thing that's being pushed into another PR for a fast follow is the functionality for libcuda direct linking. I'll merge this PR for now and then work on filing the separate PR's and backports.
